src/metabase/util/analytics.clj

```diff
@@ -0,0 +1,119 @@
+  :visibility :public)
+
+(def ^:private ^Emitter emitter
+  "An instance of a Snowplow emitter"
+  (delay
+    (let [client (-> (HttpClients/custom)
+                     (.setConnectionManager (PoolingHttpClientConnectionManager.))
+                     (.build))
+          builder (-> (ApacheHttpClientAdapter/builder)
+                      (.httpClient client)
+                      (.url (snowplow-url)))
+          adapter (.build ^ApacheHttpClientAdapter$Builder builder)
+          batch-emitter-builder (-> (BatchEmitter/builder)
+                                    (.bufferSize 1)
+                                    (.httpClientAdapter adapter))]
+      (.build ^BatchEmitter$Builder batch-emitter-builder))))
+
+(def ^:private ^Tracker tracker
+  "An instance of a Snowplow tracker"
+  (delay
+   (-> (Tracker$TrackerBuilder. ^Emitter @emitter "sp" "metabase")
+       (.build))))
+
+(defn- subject
+  "Create a Subject object for a given user ID, to be included in analytics events"
+  [user-id]
+  (-> (Subject$SubjectBuilder.)
+      (.userId (str user-id))
+      (.build)))
+
+(defn- context
+  "Common context included in every analytics event"
+  []
+  (new SelfDescribingJson
+       "iglu:com.metabase/instance/jsonschema/1-0-0"
+       {"id"             (public-settings/analytics-uuid),
+        "version"        {"tag" (:tag (public-settings/version))},
+        "token-features" (into {} (for [[token enabled?] (public-settings/token-features)]
+                                    [(name token) enabled?]))}))
+
+(defn- payload
+  "A SelfDescribingJson object containing the provided event data, which can be included as the payload for an
+  analytics event"
+  [schema version event-data]
+  (new SelfDescribingJson
+       (format "iglu:com.metabase/%s/jsonschema/%s" (name schema) version)
+       ;; Make sure keywords are converted to strings
+       (into {} (for [[k v] event-data] [(name k) (if (keyword? v) (name v) v)]))))
+
+(defn- track-schema-event
+  "Send a single analytics event to the Snowplow collector, if tracking is enabled for this MB instance"
+  [schema version user-id event-data]
+  (when (public-settings/anon-tracking-enabled)
+    (try
+      (let [^Unstructured$Builder builder (-> (. Unstructured builder)
+                                              (.eventData (payload schema version event-data))
+                                              (.customContext [(context)]))
+            ^Unstructured$Builder builder' (if user-id
+                                             (.subject builder (subject user-id))
+                                             builder)
+            ^Unstructured event (.build builder')]
+        (.track ^Tracker @tracker event))
+      (catch Throwable e
+        (log/debug e (trs "Error sending Snowplow analytics event {0}" (name (:event event-data))))))))
+
+;; Snowplow analytics interface
+
+(defmulti track-event
+  "Send a single analytics event to Snowplow"
+  (fn [event & _] event))
```

Comment:
```suggestion
  (fn [event & _] (keyword event)))
```

`defmulti` actually creates a new entry in its cache for every single dispatch value it's called with, so `dispatch` values _must_ be bounded. Wrapping the dispatch value in `keyword` enforces this. 

Otherwise you could do something evil like

```clj
(future
  (while true
    (track-event (random-string))))
```

and eventually crash the instance because it would run out of memory because the `defmulti` cache would grow forever.

Keywords are interned, so by definition they are bounded too. You'll run out of memory if you keep dynamically generating new keywords too, but that's a clearer no-no anyway and at least your multimethod won't be vulnerable